### DISCUSSION TITLE
Efficient vertex mesh updating in realtime

### DISCUSSION THREAD

**Discussion by dafteran4** [bevyengine/bevy#12751] [Q&A] [answered]
I have a soft body simulation on CPU and would like to upload changed vertex coordinates and normals every frame. I am currently doing this by storing `Handle<Mesh>`, then getting the `Mesh` asset and modifying its vertex attributes. This works, but it takes quite a performance hit. Is there some other way I could go about this?

**Comment by JMS55** [comment]
Uploading large amounts of CPU data to the GPU is expensive, there's not really a workaround. That said there might be unnecessary overhead in Bevy on top of the graphics APIs, I don't know.

**Reply by dafteran4** [reply]
Yes, I realize that. Though it seems I am orders of magnitude away from pci bandwidth limit. Anyway, I'll be looking into some other ways. I'll try to upload as little as necessary and do veretx positioning on GPU. I'm now looking into ways how to efficiently upload uniform data, similarly to how it's done in `SkinnedMesh` using `BufferVec` for updating the matrix. Are there any simpler examples for this, beside the actual skinning implementation?

**Comment by dafteran4** [comment, marked as answer]
Testing this further, I am getting much faster performance when using as few `Mesh` objects as possible. Then updating vertex attributes is fast enough. It seems my initial performance hit was more due to having a large number of `Mesh` objects.

**Reply by IceSentry** [reply]
Do you have a number for large here?

**Reply by dafteran4** [reply]
A few thousand cca, all mutated every frame. This dropped fps to about 40.
